# 1. Pengenalan

Milestone 1

Nama  : Hisham Audiputra Suharto
Batch : CODA-RMT-006

Program ini untuk fungsi ETL dengan menggunakan Webscraping

Website yang di gunakan adalah  : https://www.bhinneka.com/

Jadi Fungsi Program ini adalah untuk mengambil data dari website Bhineka Khusu Untuk Produk Storage, yang dimana yang di ambil adalah nama produknya, Harga dari produknya, cicilan bulanan untuk harga produknya dan Link spesifict untuk ke produk yang di inginkan.

# 2 .Import Library

In [11]:
import pandas as pd
from bs4 import BeautifulSoup as bs
from selenium import webdriver as wb
import time

# 3. Extract Data

## 3.1 Extract Data dari 3 page Website Bhineka Khusus untuk produk Storage

In [ ]:
# membuat drive untuk chrome
driver = wb.Chrome()
# Inisialisasi untuk variable penamping webscraping
List_Nama_Produk = []   
List_Cicilan_Produk = []
List_Harga_Produk = []
List_Link_Produk = []
# looping untuk web scarping supaya bisa pindak dari page ke page
for angka in range(1,4):
    # ubah variable page sesuai dengan looping {angka} itu akan di ganti dengan loopigan range 1 - 3 (4 - 1)
    web_link = f'https://www.bhinneka.com/jual-storage/rM1ya1P?page={angka}'
    # akses link urlnya
    driver.get(web_link)
    # delay untuk menunggu pagenya fully load memilih 10 detik karena itu cukup lama untuk pagenya fully load
    time.sleep(10)
    # mengambil source code dari pagenya 
    html = driver.page_source
    # mengformat dalam bentuk html source code dari pagenya
    soup = bs(html, 'html.parser')
    # menampung file containernya yang di mana semua element yang ingin kita ambil berada
    storage = soup.find_all('div', {'class':'o_wsale_product_grid_wrapper position-relative h-100'})
    # looping berdasarkan jumlah container
    for data in storage:
            # gunakan try and except dimana program akan berjalan dan mengambil element yang kita ingin kan
            # kalo tidak ada akan generate error dan pindah except
            try:
                # menamampung value dari element yang menyimpan nama produk
                List_Nama_Produk.append(data.find('a', {'itemprop':'name'}).get_text())
            except:
                # kalo tidak ketemu akan di gantikan jadi none (null value)
                List_Nama_Produk.append(None)
            try:
                # menamampung value dari element yang menyimpan Cicilan Produk
                List_Cicilan_Produk.append(data.find('span', {'class':'bmd-installment'}).get_text())
            except:
                # kalo tidak ketemu akan di gantikan jadi none (null value)
                List_Cicilan_Produk.append(None)
            try:
                # menamampung value dari element yang menyimpan Harga Produk
                List_Harga_Produk.append(data.find('span', {'itemprop':'price'}).get_text())
            except:
                # kalo tidak ketemu akan di gantikan jadi none (null value)
                List_Harga_Produk.append(None)
            try:
                List_Link_Produk.append('https://www.bhinneka.com/' + data.find('a', {'itemprop':'url'})['href'])
                # menamampung value dari element yang menyimpan Link_Produk
                # di karenakan waktu value href di ambil hanya ada endpoint saja tidak ada base url maka perlu di tambahkan base urlnya
            except:
                # kalo tidak ketemu akan di gantikan jadi none (null value)
                List_Link_Produk.append(None)
driver.close()

## 3.2 Masukan data data yang sudah di tampung ke dalam dataframe

In [ ]:
# Pembuatan Data Frame dengan menggunaka keys yang sesuai ex:  Data_Produk['Nama_Produk'] = List_Nama_Produk
# alasanya supaya nanti gampang di baca dan mengerti
Data_Produk = pd.DataFrame()
Data_Produk['Nama_Produk'] = List_Nama_Produk
Data_Produk['Cicilan_Produk_Perbulan'] = List_Cicilan_Produk
Data_Produk['Harga_Produk'] = List_Harga_Produk
Data_Produk['Link_Produk'] = List_Link_Produk
Data_Produk

Nama_Produk  \
0    FUJIFILM LTO-8 Data Cartridge (12TB - 30TB) wi...   
1        HP LTO5 Ultrium RW Bar Code Label Pack Q2011A   
2                             DELL EqualLogic PS6610ES   
3                          HPE MSA 2060 R0Q74A (7.2TB)   
4    HPE Nimble Storage AF40 All Flash Dual Control...   
..                                                 ...   
115     HP LTO-6 Ultrium RW Bar Code Label Pack Q2013A   
116  DELL EMC Storage SC7020F 3Ux30 Drive Storage A...   
117         HPE 3PAR 8440 2N+SW Storage (36x1.2TB SAS)   
118  LENOVO ThinkSystem DE2000H 10GBASE-T Hybrid Fl...   
119  LENOVO ThinkSystem DE2000H iSCSI Hybrid Flash ...   

         Cicilan_Produk_Perbulan  Harga_Produk  \
0        Cicilan Rp. 108.250/bln     1299000.0   
1         Cicilan Rp. 79.833/bln      958000.0   
2    Cicilan Rp. 157.712.500/bln  1892550000.0   
3     Cicilan Rp. 16.583.333/bln   199000000.0   
4    Cicilan Rp. 476.879.583/bln  5722555000.0   
..                           ...           ...   
115       Cicilan Rp. 74.542/bln      894500.0   
116  Cicilan Rp. 584.969.667/bln  7019636000.0   
117  Cicilan Rp. 414.058.292/bln  4968699500.0   
118    Cicilan Rp. 9.742.042/bln   116904500.0   
119    Cicilan Rp. 8.619.292/bln   103431500.0   

                                           Link_Produk  
0    https://www.bhinneka.com//fujifilm-lto-8-data-...  
1    https://www.bhinneka.com//hp-lto5-ultrium-rw-b...  
2    https://www.bhinneka.com//dell-equallogic-ps66...  
3    https://www.bhinneka.com//hpe-msa-2060-r0q74a-...  
4    https://www.bhinneka.com//hpe-nimble-storage-a...  
..                                                 ...  
115  https://www.bhinneka.com//hp-lto-6-ultrium-rw-...  
116  https://www.bhinneka.com//dell-emc-storage-sc7...  
117  https://www.bhinneka.com//hpe-3par-8440-2n-sw-...  
118  https://www.bhinneka.com//lenovo-thinksystem-d...  
119  https://www.bhinneka.com//lenovo-thinksystem-d...  

[120 rows x 4 columns]

## 3.3 pilih 5 data teratas dalam data frame 

In [ ]:
# mengambil data dengan menggunakan fungsi head yang dimana defaultnya mengambil 5 data teratas (untuk un-clean data)
Data_Produk.head()

Nama_Produk  \
0  FUJIFILM LTO-8 Data Cartridge (12TB - 30TB) wi...   
1      HP LTO5 Ultrium RW Bar Code Label Pack Q2011A   
2                           DELL EqualLogic PS6610ES   
3                        HPE MSA 2060 R0Q74A (7.2TB)   
4  HPE Nimble Storage AF40 All Flash Dual Control...   

       Cicilan_Produk_Perbulan  Harga_Produk  \
0      Cicilan Rp. 108.250/bln  1.299000e+06   
1       Cicilan Rp. 79.833/bln  9.580000e+05   
2  Cicilan Rp. 157.712.500/bln  1.892550e+09   
3   Cicilan Rp. 16.583.333/bln  1.990000e+08   
4  Cicilan Rp. 476.879.583/bln  5.722555e+09   

                                         Link_Produk  
0  https://www.bhinneka.com//fujifilm-lto-8-data-...  
1  https://www.bhinneka.com//hp-lto5-ultrium-rw-b...  
2  https://www.bhinneka.com//dell-equallogic-ps66...  
3  https://www.bhinneka.com//hpe-msa-2060-r0q74a-...  
4  https://www.bhinneka.com//hpe-nimble-storage-a...

## 3.4 export data mentahan ke CSV

In [32]:
Data_Produk.to_csv('Data_Produk.csv', index=False) 
# menggunakan parameter index false untuk memastikan kalo indexnya juga tidak ke export

## 3.5 Liat structure data

In [33]:
Data_Produk.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 120 entries, 0 to 119
Data columns (total 4 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   Nama_Produk              120 non-null    object
 1   Cicilan_Produk_Perbulan  120 non-null    object
 2   Harga_Produk             120 non-null    object
 3   Link_Produk              120 non-null    object
dtypes: object(4)
memory usage: 3.9+ KB


Bisa di lihat bahwa tidak ada missing value sama sekali di tiap column karena jumlahnya 120 untuk tiap kolum
jadi tidak perlu di adakan cleansing data untuk missing value

# 4. Transform Data

## 4.1 Melakukan Cleansing data pada column Cicilan_Produk_Perbulan dan juga perubahan type data

In [ ]:
# di karena kan untuk value di column Cicilan Produk Perbulan ber format 'Rp. 108.250/bln' karena harusnya column ini berbentuk int karena ini adalah harga yang berformat uang
# kita perlu memproses datanya menjadi hanya angka (menggunakan str.replace untuk format data dan astype untuk format type data column)
Data_Produk['Cicilan_Produk_Perbulan'] = Data_Produk['Cicilan_Produk_Perbulan'].str.replace('Cicilan Rp. ', '')
# mengubah value Rp.  dalam value (string) menjadi string kosong
Data_Produk['Cicilan_Produk_Perbulan'] = Data_Produk['Cicilan_Produk_Perbulan'].str.replace('/bln', '')
# mengubah value /bln  dalam value (string) menjadi string kosong
Data_Produk['Cicilan_Produk_Perbulan'] = Data_Produk['Cicilan_Produk_Perbulan'].str.replace('.', '')
# mengubah column Cicilan_Produk_Perbulan menjadi float (karena angka bisa jadi int kalo mau)
Data_Produk['Cicilan_Produk_Perbulan'] = Data_Produk['Cicilan_Produk_Perbulan'].astype(float)

## 4.2  Melakukan perubahan type data pada Harga_Produk  

In [ ]:
# di karenakan untuk harga produk sudah angka semua jadi hanya perlu merubah type column
# mengubah column Cicilan_Produk_Perbulan menjadi float (karena angka bisa jadi int kalo mau)
Data_Produk['Harga_Produk'] = Data_Produk['Harga_Produk'].astype(float)


## 4.3 pilih 5 data teratas

In [ ]:
# mengambil data dengan menggunakan fungsi head yang dimana defaultnya mengambil 5 data teratas (untuk clean data)
Data_Produk.head()

Nama_Produk  Cicilan_Produk_Perbulan  \
0  FUJIFILM LTO-8 Data Cartridge (12TB - 30TB) wi...                 108250.0   
1      HP LTO5 Ultrium RW Bar Code Label Pack Q2011A                  79833.0   
2                           DELL EqualLogic PS6610ES              157712500.0   
3                        HPE MSA 2060 R0Q74A (7.2TB)               16583333.0   
4  HPE Nimble Storage AF40 All Flash Dual Control...              476879583.0   

   Harga_Produk                                        Link_Produk  
0  1.299000e+06  https://www.bhinneka.com//fujifilm-lto-8-data-...  
1  9.580000e+05  https://www.bhinneka.com//hp-lto5-ultrium-rw-b...  
2  1.892550e+09  https://www.bhinneka.com//dell-equallogic-ps66...  
3  1.990000e+08  https://www.bhinneka.com//hpe-msa-2060-r0q74a-...  
4  5.722555e+09  https://www.bhinneka.com//hpe-nimble-storage-a...

## 4.4 Check Structure data abis di clenasing 

In [41]:
Data_Produk.info()
Data_Produk.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 120 entries, 0 to 119
Data columns (total 4 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Nama_Produk              120 non-null    object 
 1   Cicilan_Produk_Perbulan  120 non-null    float64
 2   Harga_Produk             120 non-null    float64
 3   Link_Produk              120 non-null    object 
dtypes: float64(2), object(2)
memory usage: 3.9+ KB


Cicilan_Produk_Perbulan  Harga_Produk
count             1.200000e+02  1.200000e+02
mean              7.015280e+07  8.418336e+08
std               1.867758e+08  2.241309e+09
min               4.575000e+04  5.490000e+05
25%               1.384875e+06  1.661850e+07
50%               3.588334e+06  4.306000e+07
75%               1.246563e+07  1.495875e+08
max               1.165559e+09  1.398670e+10

Bisa di lihat bahwa type data untuk column Cicilan_Produk_Perbulan dan Harga_Produk sudah berubaj jadi float dan pas di describe bisa di lihat unutk count, mean, min dll untuk data yang sudah kita ubah

## 4.4 Import table yang sudah di cleansing

In [ ]:
Data_Produk.to_csv('Data_Produk_Storage_Final.csv',index=False)
# menggunakan parameter index false untuk memastikan kalo indexnya juga tidak ke export

# 5 Load data

``` sql

-- Create Database

CREATE DATABASE Bhinneka_DB;

-- Create Table Sesuai dengan data yang kita ambil

CREATE Table Produk
(
	PK_Produk SERIAL PRIMARY KEY, 
	-- Tambahan Column PK_Produk untuk jadi primary keynya (supaya tiap data unik) dan pake serial supaya auto increment (jadi data waktu di insert langsung terisi sendiri)
	Nama_Produk TEXT,
	Cicilan_Produk_Perbulan NUMERIC,
	Harga_Produk NUMERIC,
	Link_Produk TEXT
);

-- Import data yang baru saja kita ambil ke dalam table Produk

COPY Produk(Nama_Produk, Cicilan_Produk_Perbulan, Harga_Produk, Link_Produk)
FROM 'D:\Hactive8\Phase 0\W4\Milestone\p0-coda006-rmt-m1-Gallahad12\Data_Produk_Storage_Final.csv'
DELIMITER ','
CSV HEADER;

-- Show table produk untuk memastikan kalau datanya memang sudah di import

SELECT *
from produk;

```